File: samples/grpc/restaurant-drone-deliveries-service-scala/src/main/scala/central/DroneDeliveriesServer.scala
package central

import akka.actor.typed.ActorSystem
import akka.grpc.scaladsl.ServerReflection
import akka.grpc.scaladsl.ServiceHandler
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.HttpRequest
import akka.http.scaladsl.model.HttpResponse
import central.Main.logger
import central.deliveries.proto.RestaurantDeliveriesService
import central.deliveries.proto.RestaurantDeliveriesServiceHandler
import central.drones.proto.DroneOverviewService
import central.drones.proto.DroneOverviewServiceHandler
import charging.proto.ChargingStationService
import charging.proto.ChargingStationServiceHandler

import scala.concurrent.Future
import scala.util.Failure
import scala.util.Success

object DroneDeliveriesServer {

  def start(
      interface: String,
      port: Int,
      droneOverviewService: central.drones.proto.DroneOverviewService,
      restaurantDeliveriesService: central.deliveries.proto.RestaurantDeliveriesService,
      chargingStationService: charging.proto.ChargingStationService,
      eventPullHandler: PartialFunction[HttpRequest, Future[HttpResponse]],
      eventPushHandler: PartialFunction[HttpRequest, Future[HttpResponse]])(
      implicit system: ActorSystem[_]): Unit = {
    import system.executionContext

    // #composeAndBind
    val service = ServiceHandler.concatOrNotFound(
      DroneOverviewServiceHandler.partial(droneOverviewService),
      RestaurantDeliveriesServiceHandler.partial(restaurantDeliveriesService),
      ChargingStationServiceHandler.partial(chargingStationService),
      eventPullHandler,
      eventPushHandler,
      ServerReflection.partial(
        List(
          DroneOverviewService,
          RestaurantDeliveriesService,
          ChargingStationService)))

    val bound = Http(system).newServerAt(interface, port).bind(service)
    // #composeAndBind
    bound.foreach(binding =>
      logger.info(
        "Drone event consumer listening at: {}:{}",
        binding.localAddress.getHostString,
        binding.localAddress.getPort))

    bound.onComplete {
      case Success(binding) =>
        val address = binding.localAddress
        system.log.info(
          "Drone Deliveries Service gRPC server started at {}:{}",
          address.getHostString,
          address.getPort)
      case Failure(ex) =>
        system.log.error("Failed to bind gRPC endpoint, terminating system", ex)
        system.terminate()
    }

  }

}

File: samples/grpc/restaurant-drone-deliveries-service-java/src/main/java/central/DroneDeliveriesServer.java
package central;

import akka.actor.typed.ActorSystem;
import akka.grpc.javadsl.ServerReflection;
import akka.grpc.javadsl.ServiceHandler;
import akka.http.javadsl.Http;
import akka.http.javadsl.model.HttpRequest;
import akka.http.javadsl.model.HttpResponse;
import akka.japi.function.Function;
import central.deliveries.proto.RestaurantDeliveriesService;
import central.deliveries.proto.RestaurantDeliveriesServiceHandlerFactory;
import central.drones.proto.DroneOverviewService;
import central.drones.proto.DroneOverviewServiceHandlerFactory;
import java.time.Duration;
import java.util.Arrays;
import java.util.concurrent.CompletionStage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class DroneDeliveriesServer {

  private static final Logger logger = LoggerFactory.getLogger(DroneDeliveriesServer.class);

  public static void start(
      ActorSystem<?> system,
      String host,
      int port,
      central.drones.proto.DroneOverviewService droneOverviewService,
      central.deliveries.proto.RestaurantDeliveriesService restaurantDeliveriesService,
      Function<HttpRequest, CompletionStage<HttpResponse>> deliveryEventsProducerService,
      Function<HttpRequest, CompletionStage<HttpResponse>> pushedDroneEventsHandler) {

    // #composeAndBind
    @SuppressWarnings("unchecked")
    var service =
        ServiceHandler.concatOrNotFound(
            DroneOverviewServiceHandlerFactory.create(droneOverviewService, system),
            RestaurantDeliveriesServiceHandlerFactory.create(restaurantDeliveriesService, system),
            ServerReflection.create(
                Arrays.asList(
                    DroneOverviewService.description, RestaurantDeliveriesService.description),
                system),
            deliveryEventsProducerService,
            // FIXME not last once actually partial
            pushedDroneEventsHandler);

    var bound = Http.get(system).newServerAt(host, port).bind(service);
    // #composeAndBind

    bound.whenComplete(
        (binding, error) -> {
          if (error == null) {
            logger.info(
                "Drone event consumer listening at: {}:{}",
                binding.localAddress().getHostString(),
                binding.localAddress().getPort());
            binding.addToCoordinatedShutdown(Duration.ofSeconds(3), system);
          } else {
            logger.error("Failed to bind gRPC endpoint, terminating system", error);
            system.terminate();
          }
        });
  }
}
